library/core/src/ptr/mod.rs

```diff
@@ -64,6 +65,15 @@
 //! separate allocated object), heap allocations (each allocation created by the global allocator is
 //! a separate allocated object), and `static` variables.
 //!
+//! ## A note on read vs. write invariants
+//!
+//! Both the pointer and the pointee's invariants must be satisfied when reading from a pointer,
+//! but only the pointer's invariants must be satisfied when writing. It is not considered UB to write
+//! data that violates a type's invariants so long as the type's invariants are satisfied by the time
+//! it is read, either through the raw pointer or the reference it originated from. This means a type
+//! may freely transition between valid and invalid states when being written to by raw pointers.
+//! Thus, when discussing safety, it may be useful to separately assert the validity of the pointer vs.
+//! the validity of the data it points to.
```

Comment:
This has not really been decided yet either. I am not even sure what you mean by pointer and pointee invariant here.

It is already the case that https://doc.rust-lang.org/nightly/std/ptr/fn.read.html says "src must point to a properly initialized value of type T". https://doc.rust-lang.org/nightly/std/ptr/fn.write.html does not say anything like that because it takes a `T` as argument, which already implicitly says that it must be a valid `T`.

Reply:
> Sorry, but large parts of unsafe Rust are simply still uncharted territory. :(

Haha yep, the deeper I dig the more that seems to be true. :)

> pointee invariant here.

If we decide this should go through I'll clarify that.

> which already implicitly says that it must be a valid T.

This is specifically about the `as_bytes` case, so maybe I should make the idea of breaking down a larger type into smaller pieces clearer. I want to be explicit that it is ok to write valid data that causes a type to become invalid so long as the type becomes valid again by the time it's read.

Reply:
> I want to be explicit that it is ok to write valid data that causes a type to become invalid so long as the type becomes valid again by the time it's read.

And I'm saying I think we can't way this in the official docs yet because https://github.com/rust-lang/unsafe-code-guidelines/issues/84 has not been officially answered yet.

Reply:
Gotya, so unfortunately this whole paragraph is dead? I'll just keep the small clarification if so.

Reply:
Yeah I am afraid so. :(

Reply:
Done.